### ISSUE TITLE
TabTransformer: Tabular Data Modeling Using Contextual Embeddings

### ISSUE THREAD

**Issue by masa-suke** [TeamAidemy/ds-paper-summaries#20]
Xin Huang, Ashish Khetan, Milan Cvitkovic, Zohar Karnin 2020. "TabTransformer: Tabular Data Modeling Using Contextual Embeddings" arXiv [cs.LG]. arXiv. https://arxiv.org/abs/2012.06678v1

- Self-AttentionベースのTransformerは、最先端の性能を達成するために、NLPモデルの標準コンポーネントになっている。
- Transformerによって生成された文脈埋め込みの有効性と解釈可能性もよく研究されている
- TransformersのNLPへの応用が成功したことに動機づけられ、本論文では、それらを表形式領域で適応させる。
- Transformerの適用により、ラベル無しデータの半教師あり学習が可能となり、また、多くのケースで、既存のモデルを上回る精度を達成した。

## Abstract
> We propose TabTransformer, a novel deep tabular data modeling architecture for supervised and semi-supervised learning. The TabTransformer is built upon self-attention based Transformers. The Transformer layers transform the embeddings of categorical features into robust contextual embeddings to achieve higher prediction accuracy. Through extensive experiments on fifteen publicly available datasets, we show that the TabTransformer outperforms the state-of-theart deep learning methods for tabular data by at least 1.0% on mean AUC, and matches the performance of tree-based ensemble models. Furthermore, we demonstrate that the contextual embeddings learned from TabTransformer are highly robust against both missing and noisy data features, and provide better interpretability. Lastly, for the semi-supervised setting we develop an unsupervised pre-training procedure to learn data-driven contextual embeddings, resulting in an average 2.1% AUC lift over the state-of-the-art methods.

(DeepL翻訳)
我々は、教師あり学習と半教師あり学習のための新しい深い表形式データモデリングアーキテクチャであるTabTransformerを提案する。TabTransformerは自己注意に基づくTransformerをベースに構築されている。Transformer層はカテゴリ特徴の埋め込みを頑健な文脈埋め込みに変換し、より高い予測精度を達成する。15の公開データセットに対する広範な実験を通して、我々はTabTransformerが表形式データに対する最新の深層学習手法を平均AUCで少なくとも1.0%上回り、木ベースのアンサンブルモデルの性能と一致することを示す。さらに、TabTransformerから学習した文脈埋め込みは、欠損データとノイズデータの両方の特徴に対して非常に頑健であり
、より良い解釈可能性を提供することを実証する。最後に、半教師付き設定において、我々はデータ駆動型の文脈埋め込みを学習する教師なし事前学習手順を開発し、その結果、最新の手法に対して平均2.1%のAUCリフトを達成した。

## コード
- [lucidrains/tab-transformer-pytorch (official)](https://github.com/lucidrains/tab-transformer-pytorch)
- [timeseriesAI/tsai](https://github.com/timeseriesAI/tsai/blob/main/tsai/models/TabTransformer.py)
- [jrzaurin/pytorch-widedeep](https://github.com/jrzaurin/pytorch-widedeep)
- [manujosephv/pytorch_tabular](https://github.com/manujosephv/pytorch_tabular)

## 解決した課題/先行研究との比較
Tableデータのモデル化は、大別して、GBDT等のツリーベースモデルと、多層パーセプトロン(MLP)(*)モデルの2種類があり、それぞれに課題がある。
*本論文のMLPとは、Transformerを含まない、一般的なニューラルネットワークをベースとしたモデルを示す。
- ツリーベースモデル
    - (a)ストリーミングデータの継続的な学習に適さない(再学習できない)
    - (b)最先端のSSL(半教師あり学習法,Semi-Superviced Leaning)に適さない
    - (c)欠損やノイズの多い特徴量データを処理する最先端の深層学習法はツリーベースに適用できない
- 多層パーセプトロン(MLP)モデル
    - (a)学習したモデルやEmbeddingを解釈できない
    - (b)欠損やノイズの多いデータに対してロバストではない
    - (c)半教師あり学習では、競争力のある性能を達成できない

本論文ではTransformerをテーブルデータに導入することで、上記の課題を解決し、既存のモデルを上回る精度を達成した。

## 技術のポイント
- 2種類の説明変数(連続値、カテゴリ変数)のうち、カテゴリ変数にEmbedding層とTransformer層を適用する(Fig.1)。
    - カテゴリ変数のEmbedding層は、カテゴリ変数をベクトル空間で表現する。
    - カテゴリ変数のTransformer層は、AttentionHeadを通して、前段のEmbedding層から入力された全ての埋め込みに注目した特徴ベクトルを出力する。
- ラベルなしデータを用いて、Tansformer層の事前学習(半教師あり学習)を行うことができる。
  本論文では、自然言語処理と同様の二種類の事前学習を適用できることを示している。
    - マスク言語モデリング(MLM) 
      k%の特徴をランダムに選択し、それらを欠損としてマスクし、マスク箇所を推論するための学習を事前学習として実施する。
    - 置換トークン検出(RTD)
      k%の特徴をランダムに選択し、それらをランダムな値で置き換え、置き換え箇所を推論するための学習を事前学習として実施する。
      これは、[ELECTRA](https://arxiv.org/pdf/2003.10555.pdf)に関する論文に基づいた学習手法。

<p align="center">
  <img src="https://ar5iv.labs.arxiv.org/html/2012.06678/assets/TabIllustration2.png" width="400"/>
</p>
<div align="center">Fig.1: The architecture of TabTransformer.</div>

## 評価指標
本論文では、UIC Repository、AutoML Challenge、Kaggleからの15の一般公開二値分類データセットで、教師あり学習と半教師あり学習の両方についてTabTransformerとベースラインモデルを評価し、既存モデルへの優位性を説いている。

実験は下記の項目を実施し、ほぼ全てのケースで優位な結果を示した。
- 教師あり学習シナリオにおける、TabTransformersとベースラインMLPの比較
- ノイズの多いデータと欠損値のあるデータに対する、TabTransformersとベースラインMLPの比較
  - ノイズの多いデータ
  - 欠損値のあるデータ
- 教師あり学習シナリオにおける、TabTransformersと各種モデルの比較
- 半教師あり学習シナリオにおける、TabTransformerを評価

以下、評価のサマリーを記載する。

#### 教師あり学習におけるTabTransformerと既存モデルの性能比較
通常の教師あり学習では、ニューラルネットワークベースの既存モデルを上回り、また、テーブルデータで最も高性能なモデルの１つであるGBDTとほぼ同等の性能を示した(Table 2) 。
| Model Name          | Mean AUC (%)   |
| ------------------- | :------------: |
| TabTransformer      | **82.8 ± 0.4** |
| MLP	                |   81.8 ± 0.4   |
| GBDT                | **82.9 ± 0.4** |
| Sparse MLP          |   81.4 ± 0.4   |
| Logistic Regression |   80.4 ± 0.4   |
| TabNet              |   77.1 ± 0.5   |
| VIB	                |   80.5 ± 0.4   |

Table 2教師あり学習におけるモデルの性能 評価指標は、各モデル
の15個のデータセットにおけるAUCスコアの平均±標準偏差

#### 半教師あり学習シナリオにおけるTabTransformerと既存モデルの性能比較
ラベルのないデータの数が多い場合、TabTransfomerを事前学習したモデル(TabTransformer-RTD, TabTransformer-MLM)は、他の競合他社を大幅に上回る性能を示した。
旧来の事前学習法を用いたTransformerやGBDTは全モデルの平均より性能が悪いが、TabTransformer-RTD/MLMは、ラベル付きデータが(全サンプル数 30k以上のうち)50、200、500のシナリオにおいて、平均AUCでそれぞれ1.2%、2.0%、2.1%以上、既存の手法より向上している(Table 3)。

| \#Labeled Data     | 50           | 200          | 500          |
| ------------------ | :----------: | :----------: | :----------: |
| TabTransformer-RTD |   66.6±0.6   |   70.9±0.6   | **73.1±0.6** |
| TabTransformer-MLM | **66.8±0.6** | **71.0±0.6** |   72.9±0.6   |
| MLP (ER)           |   65.6±0.6   |   69.0±0.6   |   71.0±0.6   |
| MLP (PL)	         |   65.4±0.6   |   68.8±0.6   |   71.0±0.6   |
| TabTransformer(ER) |   62.7±0.6   |   67.1±0.6   |   69.3±0.6   |
| TabTransformer(PL) |   63.6±0.6   |   67.3±0.7   |   69.3±0.6   |
| MLP (DAE)	         |   65.2±0.5   |   68.5±0.6   |   71.0±0.6   |
| GBDT (PL)	         |   56.5±0.5   |   63.1±0.6   |   66.5±0.7   |

Table 3: ラベル付きデータ点数を変えた場合の、それぞれ30K以上のデータ点を持つ8つのデータセットに対する半教師あり学習結果。評価指標は平均AUC(%)で、数値が大きいほど良い結果である。

補足：
- TabTransformer-RTD: TabTransformerを置換トークン検出(RTD)法を用いて事前学習したモデル
- TabTransformer-MLM: TabTransformerをマスク言語モデリング(MLM)法を用いて事前学習したモデル
- ER: Entropy Regularization (ER) (Grandvalet and Bengio 2006) とMLP、TabTransformer
- PL: Pseudo Labeling (PL) (Lee 2013) とMLP、TabTransformer、GBDT (Jain 2017)
- MLP (DAE): 表データに対する深いモデル用に設計された教師なし事前学習法(スワップノイズ Denoising AutoEncoder)

総じて、TabTransformerは、テーブルデータで最も高性能なモデルの１つであるGBDTに迫る精度を示し、また、ラベル付されたデータが少ないケースにおいては、半教師あり学習を適用することで、既存モデルを凌駕する性能を実現したと言える。

## 残された課題・議論
半教師あり学習時のTabTransformer評価において、特にラベルなしデータの数が少ない場合に、TabTransformer(RTD)は、ほとんどの競合を凌駕するが、改善はわずかであると述べている。
本論文のアプローチでは、ラベル無しデータ数が少ない場合、情報量の多いEmbeddingを得ることができるが、ラベルのないデータのみでは、モデル全体(特に、Fig.1 のMuliti-Layer Perceptron部)の重みを学習させることはできない。

## 重要な引用
- Arik, S. O.; and Pfister, T. 2019. TabNet: Attentive Interpretable Tabular Learning. arXiv preprint arXiv:1908.07442
URL https://arxiv.org/abs/1908.07442.
- Brunner, G.; Liu, Y.; Pascual, D.; Richter, O.; and Wattenhofer, R. 2019. On the validity of self-attention as explana
tion in transformer models. arXiv preprint arXiv:1908.04211.
- Clark, K.; Luong, M.-T.; Le, Q. V.; and Manning, C. D. 2020. ELECTRA: Pre-training Text Encoders as Discriminators Rather Than Generators. In International Conference on Learning Representations. URL https://openreview.net/forum?id=r1xMH1BtvB.
